File: Assets/Synth/Envelope.js
#pragma strict

class Envelope {
    var attack = 0.018;
    var release = 0.2;
    var current = 0.0;
    var amplifier = 1.0;

    private var delta = 0.0;

    function Envelope() {
    }

    function KeyOn() {
        delta = 1.0 / (attack * SynthConfig.kSampleRate);
    }

    function Update() {
        if (delta > 0.0) {
            current += delta;
            if (current >= 1.0) {
                current = 1.0;
                delta = -1.0 / (release * SynthConfig.kSampleRate);
            }
        } else {
            current = Mathf.Max(current + delta, 0.0);
        }
    }

    function GetLevel() {
        return current * amplifier;
    }
}

File: Assets/Synth/Envelope.cs
namespace Arpie.Synth {

struct Envelope
{
    const float Attack = 0.018f;
    const float Release = 0.2f;

    float _current;
    float _delta;

    static public Envelope Default()
      => new Envelope { _delta = 1 / (Attack * Config.SampleRate) };

    public float Run()
    {
        _current += _delta;

        if (_delta > 0 && _current >= 1)
            _delta = -1 / (Release * Config.SampleRate);

        return UnityEngine.Mathf.Clamp01(_current);
    }
}

} // namespace Arpie.Synth
